coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/services/CoinSelectionService.scala

```diff
@@ -79,8 +81,38 @@ object CoinSelectionService {
     }
   }
 
+  @tailrec
+  private def mergeOutputsPickingRec(
+      utxos: List[Utxo],
+      targetAmount: BigInt,
+      maxUtxos: Int,
+      feesPerUtxo: BigInt,
+      choosenUtxos: List[Utxo] = Nil
+  ): IO[List[Utxo]] = {
+
+    if (choosenUtxos.map(_.value).sum >= targetAmount)
+      IO.pure(choosenUtxos)
+    if (choosenUtxos.size == maxUtxos || utxos.isEmpty)
+      IO.pure(choosenUtxos)
+    else
+      mergeOutputsPickingRec(
+        utxos.tail,
+        maxUtxos,
+        feesPerUtxo,
+        if (isNearDust(utxos.head.value - feesPerUtxo))
+          choosenUtxos ::: List(utxos.head)
+        else
+          choosenUtxos
+      )
+
```

Comment:
```suggestion
final case class MergOutputState(utxo:Option[Utxo], sum:BigInt, count:Int)

private def mergeOutputsPickingRec(utxos: List[Utxo],
                                    targetAmount: BigInt,
                                    maxUtxos: Int,
                                    feesPerUtxo: BigInt,
                                  ): List[Utxo] = {

  def isStateValid(state:MergOutputState):Boolean =
    (state.count <= maxUtxos
      || state.utxo.map(_.value - feesPerUtxo).exists(isNearDust)
      || state.sum <= targetAmount)

  utxos
    .sortBy(_.value)
    .scanLeft(MergOutputState(None, 0, 0)) {
      case (state, utxo) => MergOutputState(Some(utxo), state.sum + utxo.value, state.count + 1)
    }
    .takeWhile(isStateValid)
    .collect { case MergOutputState(Some(utxo), _, _) => utxo}

```
I don't think we need a reccursive strategy here if we sort the list beforehand (see suggestion).
You can thread state by calling scanLeft.

Reply:
Using takewhile will stop immediatly if the smallest utxos are dust (and not near dust)
We could also filter them out before using this algo
If we have more than maxUtxos near dust in our account, we want to merge old utxos first.
In any case, I think we can wait to have the information about what api we really want to provide to the end user ? Then we'll be able to optimize this algo and we'll try to find a nicer way than a tailrec call

Reply:
On another axis :
- you do not need an IO as a return type, I don't see any side effect.
- `Y ::: List(X)` is very bad in term of performance, you may use `X :: Y` and reverse the list when returning

You may also factor out the escape condition and the tailrec part in sub functions.

Reply:
👍🏻

Reply:
Also you may want to correct the typo `choosen` to `chosen`

Reply:
=_=